# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGLFramebuffer.h
# Commit: ccfa767dba64
# Full Hash: ccfa767dba644dc193e0246eb8e8ff3377e8b8a5
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Client-side bindings mirror for precise CC, and merge similar codepaths. r=handyman
#   
#   * Context loss using RAII
#   * Move Program reflection Client-side
#   
# ==============================================================================

diff -r 71c122ac0ca7 -r ccfa767dba64 dom/canvas/WebGLFramebuffer.h
--- a/dom/canvas/WebGLFramebuffer.h	Wed Jan 08 22:19:14 2020 +0000
+++ b/dom/canvas/WebGLFramebuffer.h	Wed Jan 08 22:19:16 2020 +0000
@@ -8,9 +8,7 @@
 
 #include <vector>
 
-#include "mozilla/LinkedList.h"
 #include "mozilla/WeakPtr.h"
-#include "nsWrapperCache.h"
 
 #include "WebGLObjectModel.h"
 #include "WebGLRenderbuffer.h"
@@ -35,9 +33,9 @@
 struct FbAttachInfo final {
   WebGLRenderbuffer* rb = nullptr;
   WebGLTexture* tex = nullptr;
-  GLint mipLevel = 0;
-  GLint zLayer = 0;
-  GLsizei zLayerCount = 1;
+  uint32_t mipLevel = 0;
+  uint32_t zLayer = 0;
+  uint32_t zLayerCount = 1;
   bool isMultiview = false;
 };
 }  // namespace webgl
@@ -50,8 +48,8 @@
   const bool mDeferAttachment = false;
 
  private:
-  WebGLRefPtr<WebGLTexture> mTexturePtr;
-  WebGLRefPtr<WebGLRenderbuffer> mRenderbufferPtr;
+  RefPtr<WebGLTexture> mTexturePtr;
+  RefPtr<WebGLRenderbuffer> mRenderbufferPtr;
   uint32_t mTexImageLayer = 0;
   uint8_t mTexImageZLayerCount = 1;
   uint8_t mTexImageLevel = 0;
@@ -70,12 +68,9 @@
 
   ////
 
-  void Unlink() { Clear(); }
-
   bool HasAttachment() const {
     return bool(mTexturePtr) | bool(mRenderbufferPtr);
   }
-  bool IsDeleteRequested() const;
 
   void Clear();
 
@@ -97,8 +92,8 @@
 
   void DoAttachment(gl::GLContext* gl) const;
 
-  MaybeWebGLVariant GetParameter(WebGLContext* webgl, GLenum target,
-                                 GLenum attachment, GLenum pname) const;
+  Maybe<double> GetParameter(WebGLContext* webgl, GLenum attachment,
+                             GLenum pname) const;
 
   bool IsEquivalentForFeedback(const WebGLFBAttachPoint& other) const {
     if (!HasAttachment() | !other.HasAttachment()) return false;
@@ -134,11 +129,11 @@
   };
 };
 
-class WebGLFramebuffer final : public WebGLRefCountedObject<WebGLFramebuffer>,
-                               public LinkedListElement<WebGLFramebuffer>,
+class WebGLFramebuffer final : public WebGLContextBoundObject,
                                public SupportsWeakPtr<WebGLFramebuffer>,
                                public CacheInvalidator {
  public:
+  MOZ_DECLARE_REFCOUNTED_VIRTUAL_TYPENAME(WebGLFramebuffer, override)
   MOZ_DECLARE_WEAKREFERENCE_TYPENAME(WebGLFramebuffer)
 
   const GLuint mGLName;
@@ -192,18 +187,8 @@
   ////
 
  public:
-  NS_INLINE_DECL_REFCOUNTING(WebGLFramebuffer)
-
   WebGLFramebuffer(WebGLContext* webgl, GLuint fbo);
-
- private:
-  ~WebGLFramebuffer() {
-    DeleteOnce();
-    InvalidateCaches();
-  }
-
- public:
-  void Delete();
+  ~WebGLFramebuffer() override;
 
   ////
 
@@ -228,7 +213,7 @@
   void DetachRenderbuffer(const WebGLRenderbuffer* rb);
   bool ValidateAndInitAttachments(GLenum incompleteFbError) const;
   bool ValidateClearBufferType(GLenum buffer, uint32_t drawBuffer,
-                               GLenum funcType) const;
+                               webgl::AttribBaseType funcType) const;
 
   bool ValidateForColorRead(const webgl::FormatUsageInfo** out_format,
                             uint32_t* out_width, uint32_t* out_height) const;
@@ -263,13 +248,12 @@
   }
 
   FBStatus CheckFramebufferStatus() const;
-  void FramebufferAttach(GLenum attachEnum,
+  bool FramebufferAttach(GLenum attachEnum,
                          const webgl::FbAttachInfo& toAttach);
-  void DrawBuffers(const nsTArray<GLenum>& buffers);
+  void DrawBuffers(const std::vector<GLenum>& buffers);
   void ReadBuffer(GLenum attachPoint);
 
-  MaybeWebGLVariant GetAttachmentParameter(GLenum target, GLenum attachment,
-                                           GLenum pname);
+  Maybe<double> GetAttachmentParameter(GLenum attachment, GLenum pname);
 
   static void BlitFramebuffer(WebGLContext* webgl, GLint srcX0, GLint srcY0,
                               GLint srcX1, GLint srcY1, GLint dstX0,